File: scripts/admin/calculate-performance.py
#!/usr/bin/env python3
"""
Calculate Performance Metrics for All Prophets

This script:
1. Loads trained model parameters from S3
2. Runs inference on the MOST RECENT data (20, 60, 120, 240 days)
3. Calculates performance metrics (MAPE, RMSE, R², Directional Accuracy) for each window
4. Stores results in ChasingProphets-ProphetPerformanceSummary table

Note: Models are trained on historical slices but evaluated on RECENT data
"""

import boto3
import json
import pandas as pd
import numpy as np
from datetime import datetime
from decimal import Decimal
import sys
from io import StringIO

# AWS clients
dynamodb = boto3.resource('dynamodb', region_name='us-east-1')
s3 = boto3.client('s3', region_name='us-east-1')

BUCKET = 'chasingprophets-models-us-east-1'
TABLES = {
    'PROPHETS': 'ChasingProphets-Prophets',
    'MODEL_FITS': 'ChasingProphets-ModelFits',
    'DATA_SLICES': 'ChasingProphets-DataSlices',
    'DATASETS': 'ChasingProphets-Datasets',
    'PERFORMANCE': 'ChasingProphets-Performance',
    'PROPHET_PERFORMANCE_SUMMARY': 'ChasingProphets-ProphetPerformanceSummary'
}

# Evaluation windows (most recent N days)
EVALUATION_WINDOWS = [20, 60, 120, 240]

def load_from_s3(key):
    """Load file content from S3"""
    response = s3.get_object(Bucket=BUCKET, Key=key)
    return response['Body'].read().decode('utf-8')

def load_csv_from_s3(key):
    """Load CSV data from S3 into DataFrame"""
    content = load_from_s3(key)
    return pd.read_csv(StringIO(content))

def calculate_returns(prices):
    """Calculate percentage returns from prices"""
    returns = []
    for i in range(1, len(prices)):
        ret = ((prices[i] / prices[i-1]) - 1) * 100
        returns.append(ret)
    return returns

def predict_slr(parameters, return_t_minus_1):
    """Simple Linear Regression prediction"""
    slope = parameters['slope']
    intercept = parameters['intercept']
    return slope * return_t_minus_1 + intercept

def predict_mlr(parameters, return_t_minus_1, return_t_minus_2):
    """Multiple Linear Regression prediction"""
    coefficients = parameters['coefficients']
    intercept = parameters['intercept']
    input_fields = parameters.get('input_fields', ['return_t-1', 'return_t-2'])
    
    # Map input fields to values
    features = []
    for field in input_fields:
        if field == 'return_t-1':
            features.append(return_t_minus_1)
        elif field == 'return_t-2':
            features.append(return_t_minus_2)
    
    prediction = intercept
    for i, coef in enumerate(coefficients):
        if i < len(features):
            prediction += coef * features[i]
    
    return prediction

def run_inference_on_recent_data(scaffold_id, parameters, asset_id):
    """
    Run inference on the MOST RECENT data for the asset
    Returns: DataFrame with full historical data and predictions
    """
    # Asset IDs are already DJIA or SPX, no mapping needed
    folder = asset_id
    
    # Load the full CSV data
    s3_path = f'data/assets/{folder}/ohlcv_full.csv'
    try:
        df = load_csv_from_s3(s3_path)
    except Exception as e:
        print(f"    ✗ Failed to load CSV from {s3_path}: {e}")
        return []
    
    # Get price column (close)
    prices = df['close'].values
    dates = df['date'].values
    
    # Calculate returns
    returns = calculate_returns(prices)
    
    predictions = []
    
    # Need at least 3 data points for MLR (t-2, t-1, t)
    min_index = 3
    
    for i in range(min_index, len(df)):
        date = dates[i]
        actual_price = prices[i]
        previous_price = prices[i-1]
        
        # Get lagged returns
        return_t_minus_1 = returns[i-2]  # returns[i-2] corresponds to day i-1
        return_t_minus_2 = returns[i-3] if i >= 3 else 0.0
        
        # Predict next return
        try:
            if scaffold_id.lower() == 'slr':
                predicted_return = predict_slr(parameters, return_t_minus_1)
            elif scaffold_id.lower() == 'mlr':
                predicted_return = predict_mlr(parameters, return_t_minus_1, return_t_minus_2)
            else:
                if len(predictions) == 0:
                    print(f"    ✗ Unknown scaffold: {scaffold_id}")
                return []
        except Exception as e:
            if len(predictions) == 0:  # Only print error once
                print(f"    ✗ Prediction error: {e}")
            return []
        
        # Convert predicted return to price
        predicted_price = previous_price * (1 + predicted_return / 100)
        
        # Actual return for this day
        actual_return = returns[i-1]
        
        predictions.append({
            'date': date,
            'actual_price': float(actual_price),
            'predicted_price': float(predicted_price),
            'actual_return': float(actual_return),
            'predicted_return': float(predicted_return)
        })
    
    return predictions

def calculate_metrics_for_window(predictions, window_days):
    """
    Calculate performance metrics for the most recent N days
    """
    if len(predictions) < window_days:
        # Not enough data for this window
        return None
    
    # Take the most recent window_days predictions
    recent_predictions = predictions[-window_days:]
    
    actual_prices = np.array([p['actual_price'] for p in recent_predictions])
    predicted_prices = np.array([p['predicted_price'] for p in recent_predictions])
    actual_returns = np.array([p['actual_return'] for p in recent_predictions])
    predicted_returns = np.array([p['predicted_return'] for p in recent_predictions])
    
    # MAPE (Mean Absolute Percentage Error)
    mape = np.mean(np.abs((actual_prices - predicted_prices) / actual_prices)) * 100
    
    # RMSE (Root Mean Squared Error)
    rmse = np.sqrt(np.mean((actual_prices - predicted_prices) ** 2))
    
    # R² Score
    ss_res = np.sum((actual_prices - predicted_prices) ** 2)
    ss_tot = np.sum((actual_prices - np.mean(actual_prices)) ** 2)
    r2 = 1 - (ss_res / ss_tot) if ss_tot != 0 else 0
    
    # Directional Accuracy (did we predict the direction correctly?)
    actual_directions = np.sign(actual_returns)
    predicted_directions = np.sign(predicted_returns)
    directional_accuracy = np.mean(actual_directions == predicted_directions) * 100
    
    return {
        'mape': float(mape),
        'rmse': float(rmse),
        'r2': float(r2),
        'directional_accuracy': float(directional_accuracy),
        'sample_size': len(recent_predictions)
    }

def process_prophet(prophet):
    """Process a single prophet and calculate its performance across all windows"""
    prophet_id = prophet['prophetId']
    model_fit_ids = prophet.get('modelFitIds', [])
    asset_id = prophet['assetId']
    
    if not model_fit_ids or len(model_fit_ids) == 0:
        print(f"⚠️  {prophet_id}: No model fits")
        return False
    
    # Use the first model fit (prophets have one fit in our current setup)
    model_fit_id = model_fit_ids[0]
    
    print(f"Processing {prophet_id}...", end='', flush=True)
    
    try:
        # Get model fit details
        model_fits_table = dynamodb.Table(TABLES['MODEL_FITS'])
        model_fit_response = model_fits_table.get_item(Key={'modelFitId': model_fit_id})
        
        if 'Item' not in model_fit_response:
            print(f" ⚠️  Model fit not found: {model_fit_id}")
            return False
        
        model_fit = model_fit_response['Item']
        scaffold_id = model_fit['scaffoldId']
        s3_parameters_path = model_fit['modelParametersPath']
        
        # Load parameters from S3
        params_key = s3_parameters_path.replace(f's3://{BUCKET}/', '')
        params_json = load_from_s3(params_key)
        params_data = json.loads(params_json)
        # Parameters are stored directly in the JSON, not nested under 'parameters' key
        parameters = params_data if 'slope' in params_data or 'coefficients' in params_data else params_data.get('parameters', params_data)
        
        # Run inference on the FULL dataset for this asset
        all_predictions = run_inference_on_recent_data(scaffold_id, parameters, asset_id)
        
        if len(all_predictions) == 0:
            print(f" ⚠️  No predictions generated")
            return False
        
        # Calculate metrics for each evaluation window
        performance_summary_table = dynamodb.Table(TABLES['PROPHET_PERFORMANCE_SUMMARY'])
        window_metrics = {}
        
        for window_days in EVALUATION_WINDOWS:
            metrics = calculate_metrics_for_window(all_predictions, window_days)
            
            if metrics is None:
                continue
            
            window_key = f"{window_days}-day"
            window_metrics[window_key] = metrics
            
            # Store in ProphetPerformanceSummary table
            performance_summary_table.put_item(Item={
                'prophetId': prophet_id,
                'aggregationWindow': window_key,
                'mape': Decimal(str(metrics['mape'])),
                'rmse': Decimal(str(metrics['rmse'])),
                'r2': Decimal(str(metrics['r2'])),
                'directionalAccuracy': Decimal(str(metrics['directional_accuracy'])),
                'sampleSize': metrics['sample_size'],
                'lastUpdated': datetime.now().isoformat()
            })
        
        if len(window_metrics) == 0:
            print(f" ⚠️  No windows calculated")
            return False
        
        # Print summary (show metrics for each window)
        metrics_str = ", ".join([
            f"{win}: MAPE={m['mape']:.2f}%" 
            for win, m in window_metrics.items()
        ])
        print(f" ✓ {metrics_str}")
        return True
        
    except Exception as e:
        print(f" ✗ Error: {str(e)}")
        return False

def main():
    """Main execution"""
    print("📊 Calculating Performance Metrics for All Prophets\n")
    
    # Get all active prophets
    prophets_table = dynamodb.Table(TABLES['PROPHETS'])
    response = prophets_table.scan(
        FilterExpression='#status = :status',
        ExpressionAttributeNames={'#status': 'status'},
        ExpressionAttributeValues={':status': 'active'}
    )
    
    prophets = response['Items']
    
    # Handle pagination
    while 'LastEvaluatedKey' in response:
        response = prophets_table.scan(
            FilterExpression='#status = :status',
            ExpressionAttributeNames={'#status': 'status'},
            ExpressionAttributeValues={':status': 'active'},
            ExclusiveStartKey=response['LastEvaluatedKey']
        )
        prophets.extend(response['Items'])
    
    print(f"Found {len(prophets)} active prophets\n")
    
    success_count = 0
    failure_count = 0
    
    for i, prophet in enumerate(prophets, 1):
        if process_prophet(prophet):
            success_count += 1
        else:
            failure_count += 1
    
    print(f"\n{'='*60}")
    print(f"SUMMARY:")
    print(f"  ✓ Success: {success_count}")
    print(f"  ✗ Failed:  {failure_count}")
    print(f"  Total:     {len(prophets)}")
    print(f"{'='*60}")

if __name__ == '__main__':
    main()

File: scripts/admin/calculate-performance.ts
#!/usr/bin/env tsx
/**
 * Calculate Performance Metrics for All Prophets
 * 
 * This script:
 * 1. Loads all active prophets with trained models
 * 2. Runs inference on the most recent 2400 days (or max available)
 * 3. Calculates prediction errors for each data point
 * 4. Aggregates metrics over rolling windows (20d, 60d, 120d, 240d)
 * 5. Stores results in ProphetPerformanceSummary table
 * 
 * This replaces the dummy seed data with REAL performance metrics.
 */

import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { 
  DynamoDBDocumentClient, 
  ScanCommand, 
  GetCommand,
  PutCommand,
  BatchWriteCommand,
  DeleteCommand
} from '@aws-sdk/lib-dynamodb';
import { 
  S3Client, 
  GetObjectCommand 
} from '@aws-sdk/client-s3';
import { spawn } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';

const REGION = process.env.VITE_AWS_REGION || 'us-east-1';
const BUCKET = `chasingprophets-models-${REGION}`;

const ddbClient = new DynamoDBClient({ region: REGION });
const ddb = DynamoDBDocumentClient.from(ddbClient);
const s3 = new S3Client({ region: REGION });

const TABLES = {
  PROPHETS: 'ChasingProphets-Prophets',
  MODEL_FITS: 'ChasingProphets-ModelFits',
  MODEL_SCAFFOLDS: 'ChasingProphets-ModelScaffolds',
  DATA_SLICES: 'ChasingProphets-DataSlices',
  DATASETS: 'ChasingProphets-Datasets',
  ASSETS: 'ChasingProphets-Assets',
  PERFORMANCE_SUMMARY: 'ChasingProphets-ProphetPerformanceSummary'
};

const AGGREGATION_WINDOWS = [20, 60, 120, 240]; // days
const MAX_INFERENCE_DAYS = 2400;

interface Prophet {
  prophetId: string;
  prophetName: string;
  assetId: string;
  status: string;
  modelFitIds: string[];
  targetProperty: string;
}

interface ModelFit {
  modelFitId: string;
  scaffoldId: string;
  dataSliceId: string;
  trainingStatus: string;
  modelParametersPath?: string;
}

interface ModelScaffold {
  scaffoldId: string;
  name: string;
  s3RemoteInferenceScriptPath: string;
}

interface DataSlice {
  dataSliceId: string;
  datasetId: string;
  startDate: string;
  endDate: string;
}

interface Dataset {
  datasetId: string;
  assetId: string;
  s3DataPath: string;
}

interface PriceData {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

interface Prediction {
  date: string;
  actual: number;
  predicted: number;
  error: number;
  percentError: number;
}

interface PerformanceMetrics {
  mape: number;
  rmse: number;
  mae: number;
  directionalAccuracy: number;
  sampleSize: number;
}

// ============================================================================
// Data Loading
// ============================================================================

async function loadFromS3(s3Path: string): Promise<string> {
  const key = s3Path.replace(`s3://${BUCKET}/`, '');
  const command = new GetObjectCommand({ Bucket: BUCKET, Key: key });
  const response = await s3.send(command);
  if (!response.Body) throw new Error(`No data at ${s3Path}`);
  return await response.Body.transformToString();
}

async function getAllProphets(): Promise<Prophet[]> {
  const command = new ScanCommand({ TableName: TABLES.PROPHETS });
  const response = await ddb.send(command);
  return (response.Items || []) as Prophet[];
}

async function getModelFit(modelFitId: string): Promise<ModelFit | null> {
  const command = new GetCommand({
    TableName: TABLES.MODEL_FITS,
    Key: { modelFitId }
  });
  const response = await ddb.send(command);
  return (response.Item as ModelFit) || null;
}

async function getScaffold(scaffoldId: string): Promise<ModelScaffold | null> {
  const command = new GetCommand({
    TableName: TABLES.MODEL_SCAFFOLDS,
    Key: { scaffoldId }
  });
  const response = await ddb.send(command);
  return (response.Item as ModelScaffold) || null;
}

async function getDataSlice(dataSliceId: string): Promise<DataSlice | null> {
  const command = new GetCommand({
    TableName: TABLES.DATA_SLICES,
    Key: { dataSliceId }
  });
  const response = await ddb.send(command);
  return (response.Item as DataSlice) || null;
}

async function getDataset(datasetId: string): Promise<Dataset | null> {
  const command = new GetCommand({
    TableName: TABLES.DATASETS,
    Key: { datasetId }
  });
  const response = await ddb.send(command);
  return (response.Item as Dataset) || null;
}

async function loadHistoricalData(s3DataPath: string, maxDays: number): Promise<PriceData[]> {
  const csvContent = await loadFromS3(s3DataPath);
  const lines = csvContent.trim().split('\n');
  const headers = lines[0].split(',');
  
  const data: PriceData[] = [];
  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(',');
    const row: any = {};
    headers.forEach((header, idx) => {
      row[header.trim().toLowerCase()] = values[idx]?.trim();
    });
    
    data.push({
      date: row.date,
      open: parseFloat(row.open),
      high: parseFloat(row.high),
      low: parseFloat(row.low),
      close: parseFloat(row.close),
      volume: parseFloat(row.volume)
    });
  }
  
  // Sort by date ascending
  data.sort((a, b) => a.date.localeCompare(b.date));
  
  // Return most recent maxDays
  return data.slice(-maxDays);
}

// ============================================================================
// Python Inference Execution
// ============================================================================

async function runPythonInference(
  scriptPath: string,
  parametersPath: string,
  dataPath: string,
  outputPath: string
): Promise<void> {
  return new Promise((resolve, reject) => {
    // Download script and parameters to temp files
    const tempDir = '/tmp/inference';
    if (!fs.existsSync(tempDir)) {
      fs.mkdirSync(tempDir, { recursive: true });
    }
    
    const localScriptPath = path.join(tempDir, 'inference.py');
    const localParamsPath = path.join(tempDir, 'parameters.json');
    const localDataPath = path.join(tempDir, 'data.csv');
    
    Promise.all([
      loadFromS3(scriptPath).then(content => fs.writeFileSync(localScriptPath, content)),
      loadFromS3(parametersPath).then(content => fs.writeFileSync(localParamsPath, content)),
      loadFromS3(dataPath).then(content => fs.writeFileSync(localDataPath, content))
    ]).then(() => {
      const python = spawn('python3', [
        localScriptPath,
        '--parameters', localParamsPath,
        '--data', localDataPath,
        '--output', outputPath
      ]);
      
      let stdout = '';
      let stderr = '';
      
      python.stdout.on('data', (data) => { stdout += data.toString(); });
      python.stderr.on('data', (data) => { stderr += data.toString(); });
      
      python.on('close', (code) => {
        if (code === 0) {
          resolve();
        } else {
          reject(new Error(`Python inference failed: ${stderr}`));
        }
      });
    }).catch(reject);
  });
}

// ============================================================================
// Metrics Calculation
// ============================================================================

function calculateReturns(prices: PriceData[], field: string = 'close'): number[] {
  const returns: number[] = [];
  for (let i = 1; i < prices.length; i++) {
    const prevPrice = (prices[i - 1] as any)[field];
    const currentPrice = (prices[i] as any)[field];
    const percentReturn = ((currentPrice / prevPrice) - 1) * 100;
    returns.push(percentReturn);
  }
  return returns;
}

function runInferenceWithReturns(
  historicalData: PriceData[],
  parameters: any,
  targetField: string = 'close'
): Prediction[] {
  const predictions: Prediction[] = [];
  const returns = calculateReturns(historicalData, targetField);
  
  // Need at least 3 returns for MLR (2 lags + 1 to predict)
  const minReturns = 3;
  if (returns.length < minReturns) return predictions;
  
  // Extract model parameters
  const { coefficients, intercept, slope, input_fields } = parameters.parameters || parameters;
  
  // Determine model type
  const isSLR = slope !== undefined;
  const isMLR = coefficients !== undefined && Array.isArray(coefficients);
  
  for (let i = minReturns; i < historicalData.length; i++) {
    const current = historicalData[i];
    const previous = historicalData[i - 1];
    const returnIndex = i - 1;
    
    try {
      let predictedReturn: number;
      
      if (isSLR) {
        // Simple Linear Regression: y = slope * x + intercept
        const x = returns[returnIndex - 1]; // t-1 return
        predictedReturn = slope * x + intercept;
      } else if (isMLR) {
        // Multiple Linear Regression: y = sum(ci * xi) + intercept
        const features: number[] = [];
        
        if (input_fields) {
          // Use specified input fields
          for (const field of input_fields) {
            if (field === 'return_t-1' && returnIndex >= 1) {
              features.push(returns[returnIndex - 1]);
            } else if (field === 'return_t-2' && returnIndex >= 2) {
              features.push(returns[returnIndex - 2]);
            }
          }
        } else {
          // Default: use t-1 and t-2 returns
          if (returnIndex >= 1) features.push(returns[returnIndex - 1]);
          if (returnIndex >= 2) features.push(returns[returnIndex - 2]);
        }
        
        predictedReturn = intercept;
        for (let j = 0; j < coefficients.length && j < features.length; j++) {
          predictedReturn += coefficients[j] * features[j];
        }
      } else {
        throw new Error('Unknown model type');
      }
      
      // Convert predicted % return to absolute price
      const previousPrice = (previous as any)[targetField];
      const predictedPrice = previousPrice * (1 + predictedReturn / 100);
      
      const actual = (current as any)[targetField];
      const error = actual - predictedPrice;
      const percentError = Math.abs(error / actual) * 100;
      
      predictions.push({
        date: current.date,
        actual,
        predicted: predictedPrice,
        error,
        percentError
      });
    } catch (err) {
      // Skip failed predictions
      continue;
    }
  }
  
  return predictions;
}

function calculateMetrics(predictions: Prediction[]): PerformanceMetrics {
  if (predictions.length === 0) {
    return { mape: 0, rmse: 0, mae: 0, directionalAccuracy: 0, sampleSize: 0 };
  }
  
  const mape = predictions.reduce((sum, p) => sum + p.percentError, 0) / predictions.length;
  const mse = predictions.reduce((sum, p) => sum + p.error * p.error, 0) / predictions.length;
  const rmse = Math.sqrt(mse);
  const mae = predictions.reduce((sum, p) => sum + Math.abs(p.error), 0) / predictions.length;
  
  let correctDirections = 0;
  for (let i = 1; i < predictions.length; i++) {
    const actualDirection = predictions[i].actual > predictions[i - 1].actual ? 1 : -1;
    const predictedDirection = predictions[i].predicted > predictions[i - 1].predicted ? 1 : -1;
    if (actualDirection === predictedDirection) {
      correctDirections++;
    }
  }
  const directionalAccuracy = (correctDirections / (predictions.length - 1)) * 100;
  
  return {
    mape: Math.round(mape * 100) / 100,
    rmse: Math.round(rmse * 100) / 100,
    mae: Math.round(mae * 100) / 100,
    directionalAccuracy: Math.round(directionalAccuracy * 100) / 100,
    sampleSize: predictions.length
  };
}

function calculateRollingMetrics(
  predictions: Prediction[],
  windowSize: number
): PerformanceMetrics | null {
  if (predictions.length < windowSize) return null;
  
  const recentPredictions = predictions.slice(-windowSize);
  return calculateMetrics(recentPredictions);
}

// ============================================================================
// Main Execution
// ============================================================================

async function calculatePerformanceForProphet(prophet: Prophet): Promise<void> {
  console.log(`\n📊 Processing: ${prophet.prophetName} (${prophet.prophetId})`);
  
  // Skip if no model fits
  if (!prophet.modelFitIds || prophet.modelFitIds.length === 0) {
    console.log('   ⚠️  No model fits found - skipping');
    return;
  }
  
  // Get first model fit
  const modelFitId = prophet.modelFitIds[0];
  const modelFit = await getModelFit(modelFitId);
  
  if (!modelFit || modelFit.trainingStatus !== 'fit') {
    console.log(`   ⚠️  Model not trained (status: ${modelFit?.trainingStatus}) - skipping`);
    return;
  }
  
  if (!modelFit.modelParametersPath) {
    console.log('   ⚠️  No parameters path - skipping');
    return;
  }
  
  // Load scaffold
  const scaffold = await getScaffold(modelFit.scaffoldId);
  if (!scaffold) {
    console.log('   ⚠️  Scaffold not found - skipping');
    return;
  }
  
  // Load data slice
  const slice = await getDataSlice(modelFit.dataSliceId);
  if (!slice) {
    console.log('   ⚠️  Data slice not found - skipping');
    return;
  }
  
  // Load dataset
  const dataset = await getDataset(slice.datasetId);
  if (!dataset) {
    console.log('   ⚠️  Dataset not found - skipping');
    return;
  }
  
  try {
    // Load historical data (most recent 2400 days)
    console.log(`   📥 Loading historical data from ${dataset.s3DataPath}...`);
    const historicalData = await loadHistoricalData(dataset.s3DataPath, MAX_INFERENCE_DAYS);
    console.log(`   ✓ Loaded ${historicalData.length} data points`);
    
    // Load model parameters
    console.log(`   📥 Loading model parameters...`);
    const parametersContent = await loadFromS3(modelFit.modelParametersPath);
    const parameters = JSON.parse(parametersContent);
    console.log(`   ✓ Parameters loaded`);
    
    // Run inference
    console.log(`   🔮 Running inference...`);
    const predictions = runInferenceWithReturns(
      historicalData,
      parameters,
      prophet.targetProperty || 'close'
    );
    console.log(`   ✓ Generated ${predictions.length} predictions`);
    
    if (predictions.length === 0) {
      console.log('   ⚠️  No valid predictions - skipping');
      return;
    }
    
    // Calculate metrics for each aggregation window
    const performanceRecords = [];
    
    for (const windowDays of AGGREGATION_WINDOWS) {
      const windowName = `${windowDays}-day`;
      const metrics = calculateRollingMetrics(predictions, windowDays);
      
      if (metrics) {
        console.log(`   📈 ${windowName}: MAPE=${metrics.mape.toFixed(2)}%, Dir=${metrics.directionalAccuracy.toFixed(1)}%`);
        
        performanceRecords.push({
          prophetId: prophet.prophetId,
          aggregationWindow: windowName,
          mape: metrics.mape,
          rmse: metrics.rmse,
          mae: metrics.mae,
          directionalAccuracy: metrics.directionalAccuracy,
          percentileError75: metrics.mape * 1.25, // Approximate
          percentileError90: metrics.mape * 1.75, // Approximate
          sampleSize: metrics.sampleSize,
          lastUpdated: new Date().toISOString()
        });
      }
    }
    
    // Store in DynamoDB
    if (performanceRecords.length > 0) {
      console.log(`   💾 Saving ${performanceRecords.length} performance records...`);
      
      for (const record of performanceRecords) {
        await ddb.send(new PutCommand({
          TableName: TABLES.PERFORMANCE_SUMMARY,
          Item: record
        }));
      }
      
      console.log(`   ✅ Performance metrics saved`);
    }
    
  } catch (err) {
    console.error(`   ❌ Error processing prophet:`, err);
  }
}

async function main() {
  console.log('🚀 Starting Performance Calculation\n');
  console.log('═'.repeat(80));
  
  // Load all prophets
  console.log('\n📋 Loading prophets...');
  const prophets = await getAllProphets();
  console.log(`   Found ${prophets.length} total prophets`);
  
  // Filter to active prophets only
  const activeProphets = prophets.filter(p => p.status === 'active');
  console.log(`   ${activeProphets.length} active prophets to process`);
  
  // Clear existing performance summaries
  console.log('\n🧹 Clearing old performance summaries...');
  const scanCommand = new ScanCommand({ TableName: TABLES.PERFORMANCE_SUMMARY });
  const existing = await ddb.send(scanCommand);
  
  if (existing.Items && existing.Items.length > 0) {
    console.log(`   Deleting ${existing.Items.length} existing records...`);
    for (const item of existing.Items) {
      await ddb.send(new DeleteCommand({
        TableName: TABLES.PERFORMANCE_SUMMARY,
        Key: {
          prophetId: item.prophetId,
          aggregationWindow: item.aggregationWindow
        }
      }));
    }
    console.log('   ✓ Old records deleted');
  } else {
    console.log('   (No existing records)');
  }
  
  // Process each prophet
  console.log('\n' + '═'.repeat(80));
  let processed = 0;
  let succeeded = 0;
  let failed = 0;
  
  for (const prophet of activeProphets) {
    processed++;
    console.log(`\n[${processed}/${activeProphets.length}]`);
    
    try {
      await calculatePerformanceForProphet(prophet);
      succeeded++;
    } catch (err) {
      console.error(`❌ Failed:`, err);
      failed++;
    }
  }
  
  // Summary
  console.log('\n' + '═'.repeat(80));
  console.log('\n✅ Performance Calculation Complete\n');
  console.log(`   Processed: ${processed}`);
  console.log(`   Succeeded: ${succeeded}`);
  console.log(`   Failed: ${failed}`);
  console.log('');
}

main().catch(console.error);
